### ISSUE TITLE
No addRootKeyId functionality

### ISSUE THREAD

**Issue by louismorgan-linaro** [biscuit-auth/biscuit-wasm#42]
I notice in the Python library, there is a `set_root_key_id` method on the `builder` construct, but nothing like that seems to exist in this library.

Are there plans to add it? Or is it equivalent to just add the root key id as a normal fact?

**Comment by louismorgan-linaro** [comment]
I am interested, although it's probably something I would have to do in my spare time. I've limited experience with WASM, so a good place to start would be guidance on that.

For clarity, if I were to set a fact when creating the biscuit like `root_key_id=1234;` would biscuit consumers in other languages, e.g. python or rust be able to provide a function to select a key the usual way? Or does the root key id need to be set in a special way, hence the dedicated method?

**Comment by louismorgan-linaro** [comment]
Thanks very much for all the info. As you've correctly deduced, my main interest is in minting tokens although it would be nice if our front end could also perform the key selection natively.

I will certainly attempt to implement the setRootKeyId function, and depending on how that goes I will ask for more help and try to implement the key selection piece.

**Comment by louisjmorgan** [comment]
Thanks for merging the commit relating to setRootKeyId, do you have a timeline for when the next version of this package would be published to the npm registry?

I would also be keen to have a look at the key selection part. I guess your idea to use a map instead of a function would be something like:

-  a wasm binding in rust that takes a javascript object as the argument.
-  inside the rust function it creates a key selection function based on that object. 
- The javascript object would be like:

```js
{
  [key id]: {public key}
}
```

- the key selection function logic would just look up the key id in that object and return the public key if it finds a match.

Is that more or less right?

Based on reading [the wasm serde bindgen docs](https://rustwasm.github.io/wasm-bindgen/reference/arbitrary-data-with-serde.html), it seems like using a json string as the argument could be faster than a js object as well?

**Comment by pcolmer** [comment]
@divarvel Are you able to give any indication when there might be a new release of the wasm package that includes #44 ?

**Comment by divarvel** [comment]
> @divarvel Are you able to give any indication when there might be a new release of the wasm package that includes #44 ?

I'm planning a release next week

**Comment by divarvel** [comment]
https://github.com/biscuit-auth/biscuit-wasm/pull/47

**Comment by louisjmorgan** [comment]
@divarvel thanks! Is anything additional required to publish the new version to npm?

**Comment by divarvel** [comment]
> @divarvel thanks! Is anything additional required to publish the new version to npm?

nope, https://www.npmjs.com/package/@biscuit-auth/biscuit-wasm/v/0.4.1

(i gave adding key selection a try, but it's a bit more involved, i don't have anything working yet)

**Comment by louisjmorgan** [comment]
>(i gave adding key selection a try, but it's a bit more involved, i don't have anything working yet)

I'm happy to have a go if you don't have time

**Comment by divarvel** [comment]
> > (i gave adding key selection a try, but it's a bit more involved, i don't have anything working yet)
> 
> I'm happy to have a go if you don't have time

i pushed a branch with the current state of things https://github.com/biscuit-auth/biscuit-wasm/tree/key-selector

part of the changes are due to an update to `crypto`.
For now it fails at the JS / WASM boundary. Public keys are objects that live within the wasm space, and JS just manipulates pointers to it. I haven't found a way to pass map of pointers. One solution if nothing else works would be to pass a JSON object, with key ids as keys and keys as string values, but that's a bit of a hack.

**Comment by louisjmorgan** [comment]
Thanks, I will take a look this weekend if you haven't solved it by then.
